src/dachshund/candidate.rs

```diff
@@ -365,25 +365,24 @@ impl<'a, TGraph: GraphBase> Candidate<'a, TGraph> {
         visited_candidates: &mut HashSet<u64>,
     ) -> CLQResult<Vec<Self>> {
         assert!(!visited_candidates.contains(&self.checksum.unwrap()));
-        let mut tie_counts : Vec<(NodeId, usize)> = self.neighborhood
+        let tie_counts : Vec<(NodeId, usize)> = self.neighborhood
             .iter().map(|(&node_id, &edge_count)| (node_id, edge_count)).collect();
 
-        // sort by number of ties, with node_id as tie breaker for deterministic behaviour
-        // [TODO] Instead of sorting the entire list, use a min heap to keep the
-        // top nodes_to_search options.
-        tie_counts.sort_by_key(|k| (Reverse(k.1), k.0));
+        let mut h = BinaryHeap::new();
+        for (node_id, num_ties) in &tie_counts {
+            h.push((Reverse(num_ties), node_id));
+            if h.len() > num_to_search {
+                h.pop();
+            }
+        }
 
-        let mut i = 0;
         let mut expansion_candidates: Vec<Self> = Vec::new();
-        for (node_id, _num_ties) in tie_counts {
+
+        for (_num_ties, &node_id) in h.into_sorted_vec().iter() {
             let candidate = self.expand_with_node(node_id)?;
             assert!(self.checksum != candidate.checksum);
             if !visited_candidates.contains(&candidate.checksum.unwrap()) {
                 expansion_candidates.push(candidate);
-                i += 1;
```

Comment:
This doesn't have exactly the same behavior as before (because now we find the 20 best candidates and if one is a duplicate, we don't replace it, we just proceed with 19).

Practically this doesn't matter because the duplicate mechanism here never actually catches any duplicates for typical workloads. I believe this is because our duplicate protection is checking things that will are being added in epoch i against things that were processed in generation i or before (and were therefore added in generation i-1 or before).